File: rust/src/code/capacity_to_ship_packages_within_d_days.rs
struct Solution {}

impl Solution {
    pub fn ship_within_days(weights: Vec<i32>, days: i32) -> i32 {
        let mut l: i32 = weights.iter().fold(0, |acc, x| acc + x) / days;
        let mut r: i32 = weights.iter().fold(0, |acc, x| acc + x);
        let mut result: i32 = r;
        while l <= r {
            let mid: i32 = (l + r) / 2;
            if Self::can_ship(mid, &weights, days) {
                if mid.checked_sub(1).is_none() {
                    break;
                }
                result = std::cmp::min(result, mid);
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        result
    }

    fn can_ship(cap: i32, weights: &Vec<i32>, days: i32) -> bool {
        let mut accu: i32 = 0;
        let mut count: i32 = 0;
        for w in weights.iter() {
            if cap < *w {
                return false;
            }
            accu += *w;
            if accu > cap {
                count += 1;
                accu = *w;
            }
        }

        if accu > 0 {
            count += 1;
        }

        count <= days
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ship_within_days_case_1() {
        // arrange
        let weights: Vec<i32> = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
        let days: i32 = 5;
        let expected: i32 = 15;

        // act
        let actual = Solution::ship_within_days(weights, days);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_ship_within_days_case_2() {
        // arrange
        let weights: Vec<i32> = vec![3, 2, 2, 4, 1, 4];
        let days: i32 = 3;
        let expected: i32 = 6;

        // act
        let actual = Solution::ship_within_days(weights, days);

        // assert
        assert_eq!(expected, actual);
    }
}

File: python/capacity_to_ship_packages_within_d_days.py
class Solution:
    def shipWithinDays(self, weights: list[int], days: int) -> int:
        """time complexity: O(n)"""
        l: int = sum(weights) // days
        r: int = sum(weights)
        while l <= r:
            mid: int = (l + r) // 2
            if self.can_ship(mid, weights, days):
                r = mid - 1
            else:
                l = mid + 1
        return l

    def can_ship(self, cap: int, weights: list[int], days: int) -> bool:
        accu_days: int = 0
        accu: int = 0
        for w in weights:
            if w > cap:
                return False
            accu += w
            if accu > cap:
                accu_days += 1
                accu = w
        if accu > 0:
            accu_days += 1
        return accu_days <= days

    def shipWithinDays2(self, weights: list[int], days: int) -> int:
        l: int = max(weights)
        r: int = sum(weights)
        while l < r:
            mid: int = l + (r - l) // 2
            if self.feasible(mid, weights, days):
                r = mid
            else:
                l = mid + 1
        return l

    def feasible(self, n: int, weights: list[int], days: int) -> bool:
        total_days: int = 1
        accu: int = 0
        for weight in weights:
            accu += weight
            if accu > n:
                total_days += 1
                accu = weight
        return total_days <= days


def test_shipWithinDays_case_1():
    # arrange
    weights: list[int] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    days: int = 5
    expected: int = 15

    # act
    solution = Solution()
    actual = solution.shipWithinDays(weights, days)

    # assert
    assert expected == actual


def test_shipWithinDays_case_2():
    # arrange
    weights: list[int] = [3, 2, 2, 4, 1, 4]
    days: int = 3
    expected: int = 6

    # act
    solution = Solution()
    actual = solution.shipWithinDays(weights, days)

    # assert
    assert expected == actual


def test_shipWithinDays_case_3():
    # arrange
    weights: list[int] = [1, 2, 3, 1, 1]
    days: int = 4
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.shipWithinDays(weights, days)

    # assert
    assert expected == actual


def test_shipWithinDays_case_4():
    # arrange
    weights: list[int] = [1, 2, 3, 1, 1]
    days: int = 4
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.shipWithinDays2(weights, days)

    # assert
    assert expected == actual
